### DISCUSSION TITLE
How to decode raw byte array into png image?

### DISCUSSION THREAD

**Discussion by Haidar0096** [ImageMagick/ImageMagick#5630] [Help] [answered]
I have a png image, I am loading its raw bytes into memory like this:
``` 
// read image as bytes
    vector<unsigned char> png; //stores the raw pixels
    ifstream file(input, ios::binary);
    file.seekg(0, ios::end);
    png.resize(file.tellg());
    file.seekg(0, ios::beg);
    file.read((char *) &png[0], png.size());
    file.close();
``` 

Now If I create a `Magick::Image` object using the bytes from `png`, then operations on this image will fail as it is not in valid PNG format.

If I decode it using [lodepng](https://lodev.org/lodepng/), then use the decode byte array to create a `Magick::Image`, then everything works fine.

My question is: 1. Is there a built-in decoder in image magick library instead of using some external code to decode the images? And I am asking for different formats, not only PNG?

**Comment by dlemstra** [comment, marked as answer]
`file.read` doesn't guarantee that all bytes are read?

**Reply by Haidar0096** [reply]
@dlemstra 
there is no problem in reading the bytes, they are read successfully, the problem is with creating a `Magick::Image` from these bytes, it won't work if I send it the bytes directly as `new Magick::Image(width, height, map, static_cast<const MagickCore::StorageType>(type), rawBytes);`

but If I decode `rawBytes` and then create a `Magick::Image`, then subsequent operations on this `Magick::Image` work normally.

So doesn't image magick contain a built in decoder for such raw bytes array?

this is how I am doing it now:

```
    void *openImageFromBytes(unsigned width, unsigned height, const char *map, int type, const void *pixels,
                             int pixelsLength) {
        try {
            std::vector<unsigned char> image; // stores the decoded pixels
            std::vector<unsigned char> pngPixels = std::vector<unsigned char>((unsigned char *) pixels,
                                                                              (unsigned char *) pixels + pixelsLength);

            //decode
            unsigned error = lodepng::decode(image, width, height, pngPixels);

//             assert that decoded pixels info is consistent with the width and height provided
            assert(image.size() == width * height * 4);

            if (error){
                return nullptr;
            }
        return new Magick::Image(width, height, map, static_cast<const MagickCore::StorageType>(type), image);
        }
        catch (...) {
            return nullptr;
        }
    }
```

**Reply by dlemstra** [reply]
You seems to be using the constructor that expect raw pixel bytes? You probably want to use a `Blob` and use that to decode the pixel data.

**Reply by Haidar0096** [reply]
@dlemstra thanks for your help, I just reached the same result:
```
   void *openImageFromBytes(const void *pixels, int pixelsLength) {
        try {
        Magick::Blob blob(pixels, pixelsLength);
        Magick::Image *image = new Magick::Image(blob);
        image -> magick("png");
        return image;
        }
        catch (...) {
            return nullptr;
        }
    }
```

This solves the problem.

Sorry If I am asking dumb questions but I am having some problems with the documentation 😅😅.
